pkg/databases/postgresql.go

```diff
@@ -0,0 +1,303 @@
+	_, err := p.store.ExecContext(ctx, `
+		UPDATE orders 
+		SET status = $1, accural = $2, uploaded_at = $3
+		WHERE id = $4 AND user_id = $5`,
+		orderData.Status, orderData.Accrual, orderData.UploadedAt.Time, orderData.ID, orderData.UserID)
+	return err
+}
+
+func (p *Postgresql) GetOrdersList(userID int, ctx context.Context) ([]entities.OrderData, error) {
+	rows, err := p.store.QueryContext(ctx, `
+		SELECT id, user_id, order_number, status, accural, uploaded_at 
+		FROM orders 
+		WHERE user_id = $1`, userID)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	var orders []entities.OrderData
+	for rows.Next() {
+		var order entities.OrderData
+		var time time2.Time
+		if err := rows.Scan(&order.ID, &order.UserID, &order.Number, &order.Status, &order.Accrual, &time); err != nil {
+			return nil, err
+		}
+		order.UploadedAt.Time = time
+		orders = append(orders, order)
+	}
+	return orders, rows.Err()
+}
+
+func (p *Postgresql) GetUnfinishedOrdersList(ctx context.Context) ([]entities.OrderData, error) {
+	rows, err := p.store.QueryContext(ctx, `
+		SELECT id, user_id, order_number, status, accural, uploaded_at 
+		FROM orders
+		WHERE status IN ('NEW', 'PROCESSING')`)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	var orders []entities.OrderData
+	for rows.Next() {
+		var order entities.OrderData
+		err := rows.Scan(&order.ID, &order.UserID, &order.Number, &order.Status, &order.Accrual, &order.UploadedAt.Time)
+		if err != nil {
+			return nil, err
+		}
+		orders = append(orders, order)
+	}
+
+	if err = rows.Err(); err != nil {
+		return nil, err
+	}
+
+	return orders, nil
+}
+
+func (p *Postgresql) GetBalance(userID int, ctx context.Context) (entities.BalanceData, error) {
+	var balance entities.BalanceData
+
+	// Current balance
+	err := p.store.QueryRowContext(ctx, `
+		SELECT id, user_id, points 
+		FROM balances 
+		WHERE user_id = $1`, userID).Scan(&balance.ID, &balance.UserID, &balance.Current)
+	if err != nil {
+		return balance, err
+	}
+
+	//Sum of withdrawals
+	err = p.store.QueryRowContext(ctx, `
+		SELECT COALESCE(SUM(amount), 0) 
+		FROM withdrawals 
+		WHERE user_id = $1`, userID).Scan(&balance.Withdrawn)
+	if err != nil {
+		return balance, err
+	}
+
+	return balance, nil
+}
+
+func (p *Postgresql) AddToBalance(userID int, amount float64, ctx context.Context) error {
+	_, err := p.store.ExecContext(ctx, `
+		INSERT INTO balances (user_id, points) 
+		VALUES ($1, $2) 
+		ON CONFLICT (user_id) 
+		DO UPDATE SET points = balances.points + $2`,
+		userID, amount)
+	return err
+}
+
+func (p *Postgresql) WithdrawFromBalance(userID int, orderNum string, amount float64, ctx context.Context) error {
+	tx, err := p.store.BeginTx(ctx, nil)
+	if err != nil {
+		return err
+	}
+
+	// Check balance
+	var currentBalance float64
+	err = tx.QueryRowContext(ctx, `
+		SELECT points 
+		FROM balances 
+		WHERE user_id = $1`, userID).Scan(&currentBalance)
```

Comment:
```suggestion
		WHERE user_id = $1 FOR UPDATE`, userID).Scan(&currentBalance)
```
https://t.me/bushigo/36

Reply:
fix